lib/SIL/SILConstants.cpp

```diff
@@ -0,0 +1,395 @@
+/// to the user.  As such, walk the inlining location of the specified node to
+/// return the first location *outside* opaque libraries.
+static SILDebugLocation skipInternalLocations(SILDebugLocation loc) {
+  auto ds = loc.getScope();
+
+  if (!ds || loc.getLocation().getSourceLoc().isValid())
+    return loc;
+
+  // Zip through inlined call site information that came from the
+  // implementation guts of the tensor library.  We want to report the
+  // message inside the user's code, not in the guts we inlined through.
+  for (; auto ics = ds->InlinedCallSite; ds = ics) {
+    // If we found a valid inlined-into location, then we are good.
+    if (ds->Loc.getSourceLoc().isValid())
+      return SILDebugLocation(ds->Loc, ds);
+    if (SILFunction *F = ds->getInlinedFunction()) {
+      if (F->getLocation().getSourceLoc().isValid())
+        break;
+    }
+  }
+
+  if (ds->Loc.getSourceLoc().isValid())
+    return SILDebugLocation(ds->Loc, ds);
+
+  return loc;
+}
+
+/// Dig through single element aggregates, return the ultimate thing inside of
+/// it.  This is useful when dealing with integers and floats, because they
+/// are often wrapped in single-element struct wrappers.
+SymbolicValue SymbolicValue::lookThroughSingleElementAggregates() const {
+  auto result = *this;
+  while (1) {
+    if (result.getKind() != Aggregate)
+      return result;
+    auto elts = result.getAggregateValue();
+    if (elts.size() != 1)
+      return result;
+    result = elts[0];
+  }
+}
+
+/// Emits an explanatory note if there is useful information to note or if there
+/// is an interesting SourceLoc to point at.
+/// Returns true if a diagnostic was emitted.
+static bool emitNoteDiagnostic(SILInstruction *badInst, UnknownReason reason,
+                               SILLocation fallbackLoc, std::string error) {
+  auto loc = skipInternalLocations(badInst->getDebugLocation()).getLocation();
+  if (loc.isNull()) {
+    // If we have important clarifying information, make sure to emit it.
+    if (reason == UnknownReason::Default || fallbackLoc.isNull())
+      return false;
+    loc = fallbackLoc;
+  }
+
+  auto &module = badInst->getModule();
+  diagnose(module.getASTContext(), loc.getSourceLoc(),
+           diag::constexpr_unknown_reason, error)
+      .highlight(loc.getSourceRange());
+  return true;
+}
+
+/// Given that this is an 'Unknown' value, emit diagnostic notes providing
+/// context about what the problem is.
+void SymbolicValue::emitUnknownDiagnosticNotes(SILLocation fallbackLoc) {
+  auto badInst = dyn_cast<SILInstruction>(getUnknownNode());
+  if (!badInst)
+    return;
+
+  std::string error;
+  switch (getUnknownReason()) {
+  case UnknownReason::Default:
+    error = "could not fold operation";
+    break;
+  case UnknownReason::TooManyInstructions:
+    // TODO: Should pop up a level of the stack trace.
+    error = "expression is too large to evaluate at compile-time";
```

Comment:
@marcrasi  Can we reword this error? Here large seems to refer to the syntactic size of the expression in the source file. It could be a small expression but may still exceed the instruction limit e.g. if it is recursive, right? Should we reword this to something more explicit like "Exceeded instruction limit: 512  when evaluating the expression at compile time"?